### PR TITLE
CPV noise calibration

### PR DESCRIPTION
**Pull request by sevdokim**
Hello! 
CPV NoiseCalibrator and CCDB DPL service for Raw2DigitsConverterSpec is implemented in this PR.

### CODE REVIEW

**Review by kharlov — DISMISSED** [review]

File: Detectors/CPV/calib/testWorkflow/NoiseCalibratorSpec.h
Comment (shahor02) [inline Detectors/CPV/calib/testWorkflow/NoiseCalibratorSpec.h]: I see you are creating a local copy of the CCDB objects. The method I've suggested you did not work?

**Review by shahor02 — COMMENTED** [review]
Hi @sevdokim, the fullCI fails since the CPV does not get its CCDB object due to the absence of the alien-token on the build server. The PR itself is fine, but if we merge it now, it will keep breaking the fullCI. 
Is it urgent for you to have it merged? If not, I would put it on hold until the alien-token is available in the CI, hopefully, very soon.

File: Detectors/CPV/calib/testWorkflow/NoiseCalibratorSpec.h
Comment (sevdokim) [reply Detectors/CPV/calib/testWorkflow/NoiseCalibratorSpec.h]: Hi @shahor02 
Your method is working very fine, thank you! I use it in Raw2DigitsConverterSpec. 
In this noise calibration I decided to make local copy in very beginning of calibration only once and don't ask for this objects in next TFs as they will never change during noise scan. Is it ok?

**Comment by sevdokim** [comment]
Hi @shahor02 
Yes, let's wait until fullCI has alien token in there. This PR is not urgent.

File: Detectors/CPV/calib/src/NoiseCalibrator.cxx
Comment (shahor02) [inline Detectors/CPV/calib/src/NoiseCalibrator.cxx]: Why don't you use `boll badMapBool[Geometry::kNCHANNELS]={};` ? In any case, if you use `new`, you should also use `delete[]`.

File: Detectors/CPV/calib/include/CPVCalibration/NoiseCalibrator.h
Comment (shahor02) [inline Detectors/CPV/calib/include/CPVCalibration/NoiseCalibrator.h]: if you are using a class (rather than a struct) then move the datamembers to `protected:` part and have getters/setters. Also, you are assigning a to these data members arrays created with `new`, but then you never call `delete[]`. The best would be to use simply vectors (instead of pointers) or use e.g. `std::unique_ptr<std::vector<float>> mPedEfficiencies`, which you can set by external raw vector pointer by `mPedEfficiencies.reset(ptr)`.

File: Detectors/CPV/calib/testWorkflow/NoiseCalibratorSpec.h
Comment (shahor02) [reply Detectors/CPV/calib/testWorkflow/NoiseCalibratorSpec.h]: No problem to create them once, but the comment above about creating a vector with `new` and not deleting it afterwards.

**Review by shahor02 — COMMENTED** [review]
Hi @sevdokim 
OK, then let's put it on hold until the token problem is solved. Meanwhile, I had a closer look on your code, see few comments below.

File: Detectors/CPV/calib/include/CPVCalibration/NoiseCalibrator.h
Comment (sevdokim) [reply Detectors/CPV/calib/include/CPVCalibration/NoiseCalibrator.h]: Hi @shahor02 
you are absolutely right, now I use unique_ptr for those objects. Can you please check again?

File: Detectors/CPV/calib/src/NoiseCalibrator.cxx
Comment (sevdokim) [reply Detectors/CPV/calib/src/NoiseCalibrator.cxx]: Not deleting array is clearly my mistake, thanks for pointing this out.

File: Detectors/CPV/calib/src/NoiseCalibrator.cxx
Comment (shahor02) [inline Detectors/CPV/calib/src/NoiseCalibrator.cxx]: do you really need to use `at(i)` which does a range check? `[i]` is faster.

File: Detectors/CPV/calib/include/CPVCalibration/PedestalCalibrator.h
Comment (shahor02) [inline Detectors/CPV/calib/include/CPVCalibration/PedestalCalibrator.h]: this returns a copy of the vectors. Is this what you want? Usually it is enough to do e.g.
```
const std::vector<o2::ccdb::CcdbObjectInfo>&  getCcdbInfoEfficienciesVector() const { return mCcdbInfoPedEfficienciesVec; }
```

File: Detectors/CPV/calib/src/NoiseCalibrator.cxx
Comment (sevdokim) [reply Detectors/CPV/calib/src/NoiseCalibrator.cxx]: That I didn't know. Thanks for pointing!

File: Detectors/CPV/calib/include/CPVCalibration/PedestalCalibrator.h
Comment (sevdokim) [reply Detectors/CPV/calib/include/CPVCalibration/PedestalCalibrator.h]: Let me try to use const std::vector<o2::ccdb::CcdbObjectInfo>&

Comment (sevdokim) [reply Detectors/CPV/calib/include/CPVCalibration/PedestalCalibrator.h]: So the answer is following: CcdbInfoVector cannot be const as it may be changed by createObjectImage() function.

Comment (shahor02) [reply Detectors/CPV/calib/include/CPVCalibration/PedestalCalibrator.h]: ok, then you can do 
```
std::vector<o2::ccdb::CcdbObjectInfo>&  getCcdbInfoEfficienciesVector() { return mCcdbInfoPedEfficienciesVec; }
```
my point was why do you need return a copy.

Comment (sevdokim) [reply Detectors/CPV/calib/include/CPVCalibration/PedestalCalibrator.h]: Hi @shahor02 
yes, I implemented return values as references. CCDBinfo as normal references and CCDB objects as const references. Just for my understanding, what is approximate time for alien token to be imlemented into fullCI?

Comment (shahor02) [reply Detectors/CPV/calib/include/CPVCalibration/PedestalCalibrator.h]: Hi @sevdokim, unfortunately, I don't know, it is followed here: https://alice.its.cern.ch/jira/browse/O2-2760

**Comment by chiarazampolli** [comment]
Hi @shahor02 , 

Do we need to wait for the CI to be fixed (token issue) to merge? We would like to test this calibration.

Chiara

**Comment by shahor02** [comment]
Will merge after local FST check

**Comment by shahor02** [comment]
The local FST fails with 
```
[536108:internal-dpl-ccdb-backend]: [00:19:12][INFO] Fetching object for route  { CPV/CPV_Pedestals/0}
[536108:internal-dpl-ccdb-backend]: [00:19:12][ERROR] Requested resource does not exist: http://alice-ccdb.cern.ch/CPV/Calib/Pedestals/1645990691385/runNumber=0/
[536108:internal-dpl-ccdb-backend]: [00:19:12][INFO] Fetching object for route  { CPV/CPV_Gains/0}
[536108:internal-dpl-ccdb-backend]: [00:19:12][ERROR] Requested resource does not exist: http://alice-ccdb.cern.ch/CPV/Calib/Gains/1645990691385/runNumber=0/
[536108:internal-dpl-ccdb-backend]: [00:19:12][INFO] Fetching object for route  { CPV/CPV_BadMap/0}
[536108:internal-dpl-ccdb-backend]: [00:19:12][ERROR] Requested resource does not exist: http://alice-ccdb.cern.ch/CPV/Calib/BadChannelMap/1645990691385/runNumber=0/
```
@ktf, any idea why the runNumber tag is used? These objects are requested with default, run-independent ccdbParamSpec spec:
https://github.com/AliceO2Group/AliceO2/blob/3ff35a831f140f67b6add0c08281eca00fb518d9/Detectors/CPV/workflow/src/RawToDigitConverterSpec.cxx#L241-L249

**Comment by ktf** [comment]
I will check.

**Comment by ktf** [comment]
#8242 should fix the issue.

**Review by shahor02 — APPROVED** [review]
After fixing the CCDBFetcher the local FST passed, merging.

**Pull request merged**
